lessons/220/rails-app/app/controllers/devices_controller.rb

```diff
@@ -1,56 +1,34 @@
-class DevicesController < ApplicationController
-  before_action :set_device, only: %i[ show update destroy ]
+# frozen_string_literal: true
 
+class DevicesController < ApplicationController
   # GET /devices
   def index
     devices = [
-      Device.new(id: 1, uuid: "9add349c-c35c-4d32-ab0f-53da1ba40a2a", mac: "5F-33-CC-1F-43-82", firmware: "2.1.6", created_at: "2024-5-28T15:21:51.137Z", updated_at: "2024-5-28T15:21:51.137Z"),
-      Device.new(id: 2, uuid: "d2293412-36eb-46e7-9231-af7e9249fffe", mac: "E7-34-96-33-0C-4C", firmware: "1.0.3", created_at: "2024-01-28T15:20:51.137Z", updated_at: "2024-01-28T15:20:51.137Z"),
-      Device.new(id: 3, uuid: "eee58ca8-ca51-47a5-ab48-163fd0e44b77", mac: "68-93-9B-B5-33-B9", firmware: "4.3.1", created_at: "2024-8-28T15:18:21.137Z", updated_at: "2024-8-28T15:18:21.137Z")
+      Device.new(id: 1, uuid: "9add349c-c35c-4d32-ab0f-53da1ba40a2a", mac: "5F-33-CC-1F-43-82", firmware: "2.1.6", created_at: "2024-5-28T15:21:51.137Z", updated_at: "2024-5-28T15:21:51.137Z").attributes,
+      Device.new(id: 2, uuid: "d2293412-36eb-46e7-9231-af7e9249fffe", mac: "E7-34-96-33-0C-4C", firmware: "1.0.3", created_at: "2024-01-28T15:20:51.137Z", updated_at: "2024-01-28T15:20:51.137Z").attributes,
+      Device.new(id: 3, uuid: "eee58ca8-ca51-47a5-ab48-163fd0e44b77", mac: "68-93-9B-B5-33-B9", firmware: "4.3.1", created_at: "2024-8-28T15:18:21.137Z", updated_at: "2024-8-28T15:18:21.137Z").attributes
     ]
-    
-    render json: devices
-  end
 
-  # GET /devices/1
-  def show
-    render json: @device
+    build_response(devices, 200)
   end
 
   # POST /devices
   def create
-    @device = Device.new(device_params)
-    @device.uuid = SecureRandom.uuid
-
-    if @device.save
-      render json: @device, status: :created, location: @device
-    else
-      render json: @device.errors, status: :unprocessable_entity
-    end
-  end
+    device = Device.new(params.slice(:mac, :firmware))
+    device.uuid = SecureRandom.uuid
 
-  # PATCH/PUT /devices/1
-  def update
-    if @device.update(device_params)
-      render json: @device
+    if device.save
```

Comment:
when rails saving, query wrapped in a transaction, so here we have 3 queries to database

```ruby
  TRANSACTION (1.2ms)  BEGIN
  Device Create (1.7ms)  INSERT INTO "devices" ("uuid", "mac", "firmware", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5) RETURNING "id"  [["uuid", nil], ["mac", "dfdf"], ["firmware", nil], ["created_at", "2024-11-06 00:53:09.435080"], ["updated_at", "2024-11-06 00:53:09.435080"]]
  TRANSACTION (0.6ms)  COMMIT
```

this code is sensitive latency to database

I do not know how other frameworks (django,laravel) works, but they can only insertion (1 query)

we can write `device.send :create_or_update` for skip transaction
it improves performance  `Requests/sec: 6866.0961 -> 9428.1055` on my mac m1

Reply:
@simpl1g @antonputra any ideas about this?

Reply:
@ermolaev yeah, it makes sense to disable transactions if Django also is not using them. It is better to check Django before I guess. But you can at least create PR

Reply:
Let's keep it in place for now.